Binary files libupnp-1.6.6.orig/upnp/src/genlib/net/http/.httpreadwrite.c.swp and libupnp-1.6.6/upnp/src/genlib/net/http/.httpreadwrite.c.swp differ
diff -X /home/marc/dontdiff -ruN libupnp-1.6.6.orig/upnp/src/genlib/net/http/httpreadwrite.c libupnp-1.6.6/upnp/src/genlib/net/http/httpreadwrite.c
--- libupnp-1.6.6.orig/upnp/src/genlib/net/http/httpreadwrite.c	2009-12-11 09:45:30.000000000 +0100
+++ libupnp-1.6.6/upnp/src/genlib/net/http/httpreadwrite.c	2009-12-11 09:43:41.000000000 +0100
@@ -39,6 +39,7 @@
 
 #include <assert.h>
 #include <stdarg.h>
+#include <sys/poll.h>
 #ifndef UPNP_USE_BCBPP
 	#ifndef UPNP_USE_MSVCPP
 		#include <inttypes.h>
@@ -172,6 +173,11 @@
               OUT uri_type * url )
 {
     int connfd;
+    int flags;
+    struct pollfd pfd[1];
+    int connected = 0;
+    int cs = 0;
+    int timeout = 5000; // 5sec
 
     http_FixUrl( destination_url, url );
 
@@ -180,6 +186,38 @@
         return UPNP_E_OUTOF_SOCKET;
     }
 
+	if ((flags = fcntl(connfd, F_GETFL, 0)) == -1)
+	        flags = 0;
+	fcntl(connfd, F_SETFL, flags | O_NONBLOCK);
+	
+	pfd[0].fd = connfd;
+	pfd[0].events = POLLOUT;
+
+	cs = connect(connfd, ( struct sockaddr * )&url->hostport.IPv4address, sizeof( struct sockaddr_in ));
+	
+	if (cs == -1 && errno == EINPROGRESS) {
+		if (poll(pfd, 1, timeout) == 1) {
+			if (pfd[0].revents & POLLOUT == POLLOUT) {
+				connected = 1;
+				//write(2, "HTTP CONNECT connected via poll\n", 31);
+			}
+		}
+	}
+	else if (cs == 0) {
+		//write(2, "HTTP CONNECT connected without poll\n", 35);
+		connected = 1;
+	}
+
+	if (!connected) {
+		shutdown( connfd, SD_BOTH );
+	        UpnpCloseSocket( connfd );
+	       	 return UPNP_E_SOCKET_CONNECT;
+	}
+	
+	// reset old flags
+	fcntl(connfd, F_SETFL, flags);
+	
+/*
     if( connect( connfd, ( struct sockaddr * )&url->hostport.IPv4address,
                  sizeof( struct sockaddr_in ) ) == -1 ) {
 #ifdef WIN32
@@ -190,7 +228,7 @@
         UpnpCloseSocket( connfd );
         return UPNP_E_SOCKET_CONNECT;
     }
-
+*/
     return connfd;
 }
 
